net/rhp/rpc_test.go

```diff
@@ -0,0 +1,88 @@
+package rhp
+
+import (
+	"bytes"
+	"math/rand"
+	"reflect"
+	"testing"
+	"testing/quick"
+
+	"go.sia.tech/core/types"
+)
+
+func (p *HostSettings) Generate(rand *rand.Rand, size int) reflect.Value {
+	t := reflect.TypeOf(HostSettings{})
+	v := reflect.New(t)
+	for i := 0; i < v.Elem().NumField(); i++ {
+		// time.Time contains unexported fields which makes testing/quick not
+		// work so we have to have custom Generate function that skips over
+		// fields containing it
+		if t.Field(i).Type.String() == "time.Time" {
+			continue
+		}
+		elem, ok := quick.Value(t.Field(i).Type, rand)
+		if !ok {
+			return reflect.Value{}
+		}
+		v.Elem().Field(i).Set(elem)
+	}
+
+	return v
+}
+
+func TestEncoderRoundtrip(t *testing.T) {
+	tests := []types.EncoderTo{
+		&RPCFormContractRequest{},
+		&RPCRenewContractRequest{},
+		&RPCFormContractHostAdditions{},
+		&RPCRenewContractHostAdditions{},
+		&RPCContractSignatures{},
+		&RPCRenewContractRenterSignatures{},
+		&RPCLockRequest{},
+		&RPCLockResponse{},
+		&RPCReadRequest{},
+		&RPCReadResponse{},
+		&RPCSectorRootsRequest{},
+		&RPCSectorRootsResponse{},
+		&RPCWriteAction{},
+		&RPCWriteRequest{},
+		&RPCWriteMerkleProof{},
+		&RPCWriteResponse{},
+		&RPCSettingsResponse{},
+		&RPCLatestRevisionRequest{},
+		&RPCLatestRevisionResponse{},
+		&RPCSettingsRegisteredResponse{},
+		&RPCExecuteProgramRequest{},
+		&WithdrawalMessage{},
+		&PayByEphemeralAccountRequest{},
+		&PayByContractRequest{},
+		&RPCRevisionSigningResponse{},
+		&RPCAccountBalanceResponse{},
+		&RPCAccountBalanceRequest{},
+		&RPCFundAccountRequest{},
+		&RPCExecuteInstrResponse{},
+		&RPCFinalizeProgramRequest{},
+		&SettingsID{},
+		&HostSettings{},
+	}
+
+	for _, val := range tests {
+		var buf bytes.Buffer
+		e := types.NewEncoder(&buf)
+		randVal, ok := quick.Value(reflect.TypeOf(val), rand.New(rand.NewSource(0)))
+		if ok {
+			randVal.Interface().(types.EncoderTo).EncodeTo(e)
+		} else {
+			val.EncodeTo(e)
+		}
+		e.Flush()
+
+		decval := reflect.ValueOf(val)
```

Comment:
I worry about this aliasing pointers; to be safe, we should use `reflect.New(reflect.TypeOf(val))`